cli/lib/kontena/cli/stacks/yaml/reader.rb

```diff
@@ -77,92 +84,150 @@ def internals_interpolated_yaml
               warnings: false
             )
           )
-        ) || {}
+        )
       rescue Psych::SyntaxError => ex
         raise ex, "Error while parsing #{file} : #{ex.message}"
       end
 
+      # Uses variable interpolation, prompts as needed, liquid interpolation
+      #
+      # @return [Hash] the most commplete stack parsing outcome
       def fully_interpolated_yaml
         return @fully_interpolated_yaml if @fully_interpolated_yaml
-        vars = variables.to_h(values_only: true)
         @fully_interpolated_yaml = ::YAML.safe_load(
           replace_dollar_dollars(
             interpolate(
               interpolate_liquid(
                 raw_content,
-                vars
+                variable_values
               ),
               use_opto: true,
               raise_on_unknown: true
             )
           )
-        ) || {}
+        )
       rescue Psych::SyntaxError => ex
         raise ex, "Error while parsing #{file} : #{ex.message}"
       end
 
+      # The YAML file raw content
+      def raw_content
+        loader.content
+      end
+
+      # @return [Hash] with zero interpolation/processing. Will mostly fail
       def raw_yaml
-        @raw_yaml ||= ::YAML.safe_load(raw_content) || {}
+        loader.yaml
       end
 
+      # Creates an opto option definition compatible hash from the #default_envs hash
+      # @return [Hash]
+      def default_envs_to_options
+        default_envs.each_with_object({}) { |env, obj| obj[env[0]] = { type: :string, value: env[1] } }
+      end
+
+      # Accessor to the Opto variable handler
       # @return [Opto::Group]
       def variables
-        return @variables if @variables
-        @variables = ::Opto::Group.new(
-          (internals_interpolated_yaml['variables'] || {}).merge(default_envs.each_with_object({}) { |env, obj| obj[env[0]] = { type: :string, value: env[1] } }),
-          defaults: {
-            from: :env,
-            to: :env
-          }
+        @variables ||= ::Opto::Group.new(
+          internals_interpolated_yaml.fetch('variables', {}).merge(default_envs_to_options),
+          defaults: { from: :env }
```

Comment:
This is a really small change, but does it also have some other side-effects? I don't know myself... I'd rather have a separate one-liner PR for this change than having to think about it here and now before being able to merge this :confused:

Reply:
Confirmed to fix the issue, though:

```
$ redis | grep PEERS
    - PEERS=
$ kontena service show deptest-bar/redis | grep PEERS
    - PEERS=foo=redis://$STACK-foo:6379/

$ kontena stack upgrade deptest test/deptest.yaml 
...

$ kontena service show deptest-foo/redis | grep PEERS
    - PEERS=
```

Reply:
Previously it turned this:

```yaml
variables:
  foo:
    type: string
    from:
      prompt: Foo?
```

to this:

```yaml
variables:
  foo:
    type: string
    from:
      env:  foo
      prompt: Foo?
    to:
      env: foo
```

The `env: foo` has no effect. The interpolator does not even read env variables any more and everything has to be declared:

```ruby
                opt = variables.option(var)
                if opt.nil?
                  raise RuntimeError, "Undeclared variable '#{var}' in #{file}:#{line_num} -- #{row}" if raise_on_unknown
                else
                  val = opt.value
                end
```

Someone could have this:

```yaml
variables:
  foo:
    type: string
    value: hello
  bar:
    type: string
    from:
      env:  foo
```

And it would break after this.

Reply:
Reverted, will pr it separately after this is merged.